### PR TITLE
GEOPY-1494: Python 3.11 (and 3.12) Support - drop Python 3.9

### PR DESCRIPTION
**Pull request by domfournier**
**GEOPY-1494 - Python 3.11 (and 3.12) Support**

Include changes from #551

### CODE REVIEW

**Comment by domfournier** [comment]
@sebhmg Not to be merged in, but same PR will go on https://github.com/SamuelBradley/geoh5py/pull/1

File: tests/ui_json_test.py [hunk 1]
```diff
@@ -540,7 +540,7 @@ def test_input_file(tmp_path: Path):
     ):
         InputFile.read_ui_json("somefile.json")
 
-    with pytest.raises(TypeError, match="expected str, bytes or os.PathLike object"):
+    with pytest.raises(TypeError):
```
Comment (sebhmg) [inline tests/ui_json_test.py:543]: **question:** why removing the test of the error message?

`match` can take a regex if it helps to have the test more resistant to minor changes in the message text

File: pyproject.toml [hunk 2]
```diff
@@ -37,11 +37,13 @@ include = [
 ]
 
 [tool.poetry.dependencies]
-python = "^3.9,<3.11"
-
+python = "^3.9,<3.13"
+Pillow = "10.1.0"
```
Comment (sebhmg) [inline pyproject.toml:41]: **suggestion:** first merge release branch to dev, to have Pillow "~10.3.0"

**Review by sebhmg — APPROVED** [review]

File: pyproject.toml (hunk 2 above)
Comment (sebhmg) [reply pyproject.toml:41]: or merge can happen after, but probably good to have it now and make sure tests are passing fine with newer python versions

File: tests/ui_json_test.py (hunk 1 above)
Comment (domfournier) [reply tests/ui_json_test.py:543]: Looks like the error message returned by pathlib in version 3.12 has changed substantially.

Comment (sebhmg) [reply tests/ui_json_test.py:543]: ho, if the message is not coming from the code of the repo itself, checking for the error text is less relevant. As long as it captures a TypeError

**Comment by sebhmg** [comment]
also another couple of changes please:
* 3.9 -> 3.10 in `.github/workflows/static-analysis.yml`
* `--py310-plus` for pyupgrade in `.pre-commit-config.yaml`

**Comment by domfournier** [comment]
To be continued on #551

**Pull request closed without merging**